server/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/IndexModeStatsActionType.java

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.action.admin.cluster.node.stats;
+
+import org.apache.lucene.store.AlreadyClosedException;
+import org.elasticsearch.action.ActionType;
+import org.elasticsearch.action.FailedNodeException;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.nodes.BaseNodeResponse;
+import org.elasticsearch.action.support.nodes.BaseNodesRequest;
+import org.elasticsearch.action.support.nodes.BaseNodesResponse;
+import org.elasticsearch.action.support.nodes.TransportNodesAction;
+import org.elasticsearch.cluster.ClusterName;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.index.IndexMode;
+import org.elasticsearch.index.IndexService;
+import org.elasticsearch.index.shard.DocsStats;
+import org.elasticsearch.index.shard.IllegalIndexShardStateException;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.injection.guice.Inject;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportRequest;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.EnumMap;
+import java.util.List;
+import java.util.Map;
+
+public final class IndexModeStatsActionType extends ActionType<IndexModeStatsActionType.StatsResponse> {
+    public static final IndexModeStatsActionType TYPE = new IndexModeStatsActionType();
+
+    private IndexModeStatsActionType() {
+        super("cluster:monitor/nodes/index_mode_stats");
+    }
+
+    public static class StatsRequest extends BaseNodesRequest {
+        public StatsRequest(String[] nodesIds) {
+            super(nodesIds);
+        }
+
+        public StatsRequest(DiscoveryNode... concreteNodes) {
+            super(concreteNodes);
+        }
+    }
+
+    public static class IndexStats implements Writeable {
+        private long numDocs;
```

Comment:
make fields final?

Reply:
Fields need to be non-final for the `add` method, which combines stats across shards and nodes.